BeatSaberTheater/Services/AuthTokenService.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using BeatSaberTheater.Util;
+using BS_Utils.Utilities;
+using BSTheater.Shared.Models.Authentication.Responses;
+using BSTheater.Shared.Services;
+using Newtonsoft.Json.Linq;
+using UnityEngine;
+using Zenject;
+
+namespace BeatSaberTheater.Services;
+
+public class AuthTokenService : IInitializable, IDisposable
+{
+    public string CurrentAuthToken = string.Empty;
+
+    private readonly LoggingService _loggingService;
+    private IPlatformUserModel? _platformUserModel;
+    private CancellationTokenSource? _refreshTokenCts;
+    private DateTime _tokenExpiryUtc = DateTime.MinValue;
+
+    public AuthTokenService(LoggingService loggingService)
+    {
+        _loggingService = loggingService;
+    }
+
+    public void Initialize()
+    {
+        BSEvents.lateMenuSceneLoadedFresh += OnMenuSceneLoadedFresh;
+    }
+
+    private void OnMenuSceneLoadedFresh(ScenesTransitionSetupDataSO? scenesTransition)
+    {
+        Task.Run(RefreshAuthToken);
+    }
+
+    private async Task RefreshAuthToken()
+    {
+        _loggingService.Debug("Attempting rotating auth token");
+
+        if (_platformUserModel is null)
+        {
+            var platformUserModel = Resources
+                .FindObjectsOfTypeAll<PlatformLeaderboardsModel>()
+                .Select(l => l._platformUserModel)
+                .Last(x => x != null);
+            _platformUserModel = platformUserModel;
+        }
+
+        var platformUserToken = await _platformUserModel.GetUserAuthToken();
+        _loggingService.Debug("PLATFORM USER: " + platformUserToken.validPlatformEnvironment);
+        _loggingService.Debug("PLATFORM USER TOKEN: " + platformUserToken.token);
+        var token = platformUserToken.token;
+
+        var authTokenResponse = await RetrieveUserToken(token);
+        CurrentAuthToken = authTokenResponse.Token;
+
+        // Parse JWT expiration (exp) claim if present
+        _tokenExpiryUtc = ParseJwtExpiry(CurrentAuthToken) ?? DateTime.UtcNow.AddMinutes(55); // Fallback to 55 minutes if expiry could not be parsed from JWT
```

Comment:
May want to log debug / warning in this case. This should not be happening 😛 and should indicate either a parsing failure on the client end, of a backend failure where we're issueing incomplete tokens

Reply:
Yeah, that should be handled in the other exception handler